blockprod/src/detail/job_manager.rs

```diff
@@ -0,0 +1,358 @@
+            let job_key = JobKey::new(current_tip_id);
+
+            if jobs.contains_key(&job_key) {
+                if let Err(e) = result_sender.send(Err(JobManagerError::JobAlreadyExists)) {
+                    log::info!("Error sending new job exists error: {e:?}");
+                }
+            } else {
+                jobs.insert(job_key.clone(), JobHandle { cancel_sender });
+
+                if let Err(e) = result_sender.send(Ok(job_key)) {
+                    log::info!("Error sending new job event: {e:?}");
+                }
+            }
+        }
+    }
+
+    pub async fn stop_job(&mut self, job_key: Option<JobKey>) -> Result<usize, JobManagerError> {
+        let (result_sender, result_receiver) = oneshot::channel();
+
+        if self.stop_job_sender.send((job_key, result_sender)).is_err() {
+            return Err(JobManagerError::FailedToSendStopJobEvent);
+        }
+
+        tokio::select! {
+            result = result_receiver => match result {
+                Ok(v) => Ok(v),
+                Err(_) => Err(JobManagerError::FailedToStopJobs),
+            }
+        }
+    }
+
+    fn stop_job_handler(
+        jobs: &mut BTreeMap<JobKey, JobHandle>,
+        event: Option<(Option<JobKey>, oneshot::Sender<usize>)>,
+    ) {
+        if let Some((job_key, result_sender)) = event {
+            let mut stop_jobs = Vec::new();
+
+            match job_key {
+                Some(job_key) => {
+                    if let Some(job_handle) = jobs.remove(&job_key) {
+                        stop_jobs.push((job_key, job_handle));
+                    }
+                }
+                None => {
+                    while let Some((job_key, job_handle)) = jobs.pop_first() {
+                        stop_jobs.push((job_key, job_handle));
+                    }
+
+                    log::info!("Cancelling {} jobs", stop_jobs.len());
+                }
+            }
+
+            let stop_count = stop_jobs.len();
+
+            for (job_key, job_handle) in stop_jobs.drain(..) {
+                let _ = job_handle.cancel_sender.send(());
+                log::info!("Stopped mining job for tip {}", job_key.current_tip_id());
+            }
+
+            if let Err(e) = result_sender.send(stop_count) {
+                log::info!("Error sending stop jobs count: {e:?}");
+            }
+        }
+    }
+
+    fn shutdown_handler(
+        jobs: &mut BTreeMap<JobKey, JobHandle>,
+        event: Option<oneshot::Sender<usize>>,
+    ) {
+        if let Some(result_sender) = event {
+            log::info!("Stopping job manager");
+            Self::stop_job_handler(jobs, Some((None, result_sender)));
+        }
+    }
+}
+
+impl Drop for JobManager {
+    fn drop(&mut self) {
+        let (result_sender, result_receiver) = oneshot::channel();
+
+        if let Err(e) = self.shutdown_sender.send(result_sender) {
+            log::info!("Error sending shutdown during job manager drop: {e:?}");
```

Comment:
error

Reply:
Yep, you're right. I've updated everywhere where `info!` should have been `error!`